.github/workflows/sel4test-hw.yml

```
@@ -0,0 +1,124 @@
+# Copyright 2021, Proofcraft Pty Ltd
+#
+# SPDX-License-Identifier: BSD-2-Clause
+
+# sel4test hardware builds and runs
+#
+# See sel4test-hw/builds.yml in the repo seL4/ci-actions for configs.
+
+name: seL4Test-HW
+
+on:
+  workflow_call:
+
+# intended to run on
+#  pull_request_target:
+#    types: [opened, reopened, synchronize, labeled]
+# needs PR target for secrets access; guard by requiring label
+
+# downgrade permissions to read-only as you would have in a standard PR action
+permissions:
+  contents: read
+
+jobs:
+  code:
+    name: Freeze Code
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'pull_request_target' &&
+              github.event.action != 'labeled' &&
+              (contains(github.event.pull_request.labels.*.name, 'hw-build') ||
+               contains(github.event.pull_request.labels.*.name, 'hw-test')) ||
+            github.event_name == 'pull_request_target' &&
+              github.event.action == 'labeled' &&
+              (github.event.label.name == 'hw-build' ||
+               github.event.label.name == 'hw-test') }}
+    outputs:
+      xml: ${{ steps.repo.outputs.xml }}
+    steps:
+    - id: repo
+      uses: seL4/ci-actions/repo-checkout@master
+      with:
+        manifest_repo: sel4test-manifest
+        manifest: master.xml
+        sha: ${{ github.event.pull_request.head.sha }}
+
+  hw-build:
+    name: HW Build
+    runs-on: ubuntu-latest
+    needs: code
+    # To reduce the load (especiually on the machine queue) we cancel any older
+    # runs of this workflow for the current PR.
+    concurrency:
+      group: ${{ github.workflow }}-sel4test-pr-build-${{ github.event.number }}
+      cancel-in-progress: true
+    strategy:
+      fail-fast: false
+      matrix:
+        # There is no "rv32imac" hardware yet.
+        march: [armv7a, armv8a, nehalem, rv64imac]
+        compiler: [gcc, clang]
+    steps:
+    - name: Build
+      uses: seL4/ci-actions/sel4test-hw@master
+      with:
+        xml: ${{ needs.code.outputs.xml }}
+        march: ${{ matrix.march }}
+        compiler: ${{ matrix.compiler }}
+    - name: Upload images
+      uses: actions/upload-artifact@v4
+      with:
+        name: images-${{ matrix.march }}-${{ matrix.compiler }}
+        path: '*-images.tar.gz'
+    - name: Upload kernel.elf files
+      uses: actions/upload-artifact@v4
+      with:
+        name: kernel.elf-${{ matrix.march }}-${{ matrix.compiler }}
+        path: '*-kernel.elf'
+
+  the_matrix:
+    name: Matrix
+    needs: hw-build
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.matrix.outputs.matrix }}
+    steps:
+    - id: matrix
+      uses: seL4/ci-actions/sel4test-hw-matrix@master
+
+  hw-run:
+    name: HW Run
+    runs-on: ubuntu-latest
+    needs: the_matrix
+    if: ${{ github.repository_owner == 'seL4' &&
```

Since we have filtered almost everything in the `code` job, checking this seems enough:
```suggestion
    if: ${{ github.repository_owner == 'seL4' &&
            contains(github.event.pull_request.labels.*.name, 'hw-test')  }}
```

This is the context for a "labeld" event. the new label is part of the array then already.
```
"github": {
  "event_name": "pull_request_target",
  "event": {
    "action": "labeled",
    "label": {
      ...
      "name": "hw-test",
      ...
    },
    "number": 140,
    "pull_request": {
      ...
      "labels": [
        {
          ...
          "name": "hw-test",
          ...          
        }
      ],
...
```

---

Are you sure the label wasn't in there before from being set previously? Because I remember doing exactly this debug print and had to specifically distinguish the two cases because in one of them it wasn't in the list.

Maybe they changed this in the meantime to simplify this case. It'd be nice to know for sure what we can rely on. Will have a look at what the docs say.

---

The [docs say](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=labeled#pull_request) that a labeled event must have a `pull_request` field which must have a `labels` field. I guess the only remaining question is if we are guaranteed that the new label is part of that `labels` field. If yes, we really can simplify.

---

> Are you sure the label wasn't in there before from being set previously? 

Yes, I have trief this a couple of times as I was wondering exactly about that. Not sure if there is a guarantee, tough.

---

I figured out why it is they way it is: if the event is `labeled` the user might be adding a label that is a different label, e.g. `docs`. In that case we don't want to re-trigger the test if it is already running (and the label `hw-test` would already be in the list of labels, because it's already been set before).

---

True,  but this is properly filtered-out by the job `code` already - and I want to keep this part. So, when we arrive here in the job `hw-run`, it can only have happened due to a push or a new "hw_build"/"hw_test" label. So we can actually simply check if we have the "hw-test" set here.

---

I don't think that's good. Why make assumptions if we can just test what we need to test? If we ever refactor or move this it'll work fine the way it is now, but not if we depend on context info from other jobs.

---

Well, there are pros and cons. We depend on the build step anyway at the moment. In a bigger picture, it would be nicer if a caller would decide when to invoke which step, but I have not found nice was for this. So let's keep it as it is for now.